Generate the interop with harfbuzz

mattleibow opened the pull request:
**Description of Change**

The interop layer for harfbuzz was manual, so switch to the generator.

**Bugs Fixed**

<!-- Provide links to issues here. Ensure that a GitHub issue was created for your feature or bug fix before sending PR. -->

- Related to issue #

**API Changes**

<!-- REPLACE THIS COMMENT
List all API changes here (or just put None), example:

Added: 
 
- `string Class.Property { get; set; }`
- `void Class.Method();`

Changed:

 - `object Cell.OldPropertyName => object Cell.NewPropertyName`
 
-->

**Behavioral Changes**

<!-- Describe any non-bug related behavioral changes that may change how users app behaves when upgrading to this version of the codebase. -->

**PR Checklist**

- [ ] Has tests (if omitted, state reason in description)
- [ ] Rebased on top of master at time of PR
- [ ] Changes adhere to coding standard
- [ ] Updated documentation

Gillibald commented on binding/HarfBuzzSharp.Shared/Buffer.cs:
```diff
@@ -235,19 +235,21 @@ public void AddCodepoints (IntPtr text, int textLength, int itemOffset, int item
+			return new ReadOnlySpan<GlyphInfo> (infoPtrs, (int)length);
 		}
 
 		public unsafe ReadOnlySpan<GlyphPosition> GetGlyphPositionSpan ()
```
Hopefully we can change the to a writable Span<T> soon

mattleibow replied:
I think we can, but maybe not in this PR. I see 2.7.2 is out, so we can break with 2.7.x.

Gillibald commented:
What do we need to get this finished?

mattleibow commented:
I am going to run the build again to see how things are, but I think it was just a matter of fixing up a few enum names and all that.

I also wanted to get the v2.80.2 release out first, and then do this in v2.80.3 right after. Just in case I break something :)

I was really hoping to get v2.80.2 out by Wednesday, but we had to move form the old package signing system to the new one, and that only worked with beautiful, clean, multi-targeting projects on Windows... Seems SkiaSharp builds some parts on this foreign OS called "macOS" and "Linux"... It also seems to have a slightly more complex NuGet package... 😁 But, I think all is good now and I just finished writing the new pipelines/jobs.

mattleibow commented:
# API diff: HarfBuzzSharp.dll

## HarfBuzzSharp.dll

### Namespace HarfBuzzSharp

#### Type Changed: HarfBuzzSharp.Feature

Added interface:

```csharp
System.IEquatable<Feature>
```

Added methods:

```csharp
public virtual bool Equals (Feature obj);
public override bool Equals (object obj);
public override int GetHashCode ();
public static bool op_Equality (Feature left, Feature right);
public static bool op_Inequality (Feature left, Feature right);
```


#### Type Changed: HarfBuzzSharp.FontExtents

Added interface:

```csharp
System.IEquatable<FontExtents>
```

Added methods:

```csharp
public virtual bool Equals (FontExtents obj);
public override bool Equals (object obj);
public override int GetHashCode ();
public static bool op_Equality (FontExtents left, FontExtents right);
public static bool op_Inequality (FontExtents left, FontExtents right);
```


#### Type Changed: HarfBuzzSharp.GlyphExtents

Added interface:

```csharp
System.IEquatable<GlyphExtents>
```

Added methods:

```csharp
public virtual bool Equals (GlyphExtents obj);
public override bool Equals (object obj);
public override int GetHashCode ();
public static bool op_Equality (GlyphExtents left, GlyphExtents right);
public static bool op_Inequality (GlyphExtents left, GlyphExtents right);
```


#### Type Changed: HarfBuzzSharp.GlyphInfo

Added interface:

```csharp
System.IEquatable<GlyphInfo>
```

Added properties:

```csharp
public int Var1 { get; set; }
public int Var2 { get; set; }
```

Added methods:

```csharp
public virtual bool Equals (GlyphInfo obj);
public override bool Equals (object obj);
public override int GetHashCode ();
public static bool op_Equality (GlyphInfo left, GlyphInfo right);
public static bool op_Inequality (GlyphInfo left, GlyphInfo right);
```


#### Type Changed: HarfBuzzSharp.GlyphPosition

Added interface:

```csharp
System.IEquatable<GlyphPosition>
```

Added methods:

```csharp
public virtual bool Equals (GlyphPosition obj);
public override bool Equals (object obj);
public override int GetHashCode ();
public static bool op_Equality (GlyphPosition left, GlyphPosition right);
public static bool op_Inequality (GlyphPosition left, GlyphPosition right);
```


#### Type Changed: HarfBuzzSharp.MemoryMode

Removed value:

```csharp
ReadOnlyMayMakeWriteable = 3,
```

Added value:

```csharp
ReadonlyMayMakeWritable = 3,
```


#### Type Changed: HarfBuzzSharp.OpenTypeMetricsTag

Removed values:

```csharp
SubScriptEmXOffset = 1935833199,
SubScriptEmXSize = 1935833203,
SubScriptEmYOffset = 1935833455,
SubScriptEmYSize = 1935833459,
SuperScriptEmXOffset = 1936750703,
SuperScriptEmXSize = 1936750707,
SuperScriptEmYOffset = 1936750959,
SuperScriptEmYSize = 1936750963,
```

Added values:

```csharp
HbOtMetricsTagMaxValue = 2147483647,
SubscriptEmXOffset = 1935833199,
SubscriptEmXSize = 1935833203,
SubscriptEmYOffset = 1935833455,
SubscriptEmYSize = 1935833459,
SuperscriptEmXOffset = 1936750703,
SuperscriptEmXSize = 1936750707,
SuperscriptEmYOffset = 1936750959,
SuperscriptEmYSize = 1936750963,
```


#### Type Changed: HarfBuzzSharp.UnicodeCombiningClass

Removed values:

```csharp
CCC10 = 10,
CCC103 = 103,
CCC107 = 107,
CCC11 = 11,
CCC118 = 118,
CCC12 = 12,
CCC122 = 122,
CCC129 = 129,
CCC13 = 13,
CCC130 = 130,
CCC133 = 132,
CCC14 = 14,
CCC15 = 15,
CCC16 = 16,
CCC17 = 17,
CCC18 = 18,
CCC19 = 19,
CCC20 = 20,
CCC21 = 21,
CCC22 = 22,
CCC23 = 23,
CCC24 = 24,
CCC25 = 25,
CCC26 = 26,
CCC27 = 27,
CCC28 = 28,
CCC29 = 29,
CCC30 = 30,
CCC31 = 31,
CCC32 = 32,
CCC33 = 33,
CCC34 = 34,
CCC35 = 35,
CCC36 = 36,
CCC84 = 84,
CCC91 = 91,
```

Added values:

```csharp
Ccc10 = 10,
Ccc103 = 103,
Ccc107 = 107,
Ccc11 = 11,
Ccc118 = 118,
Ccc12 = 12,
Ccc122 = 122,
Ccc129 = 129,
Ccc13 = 13,
Ccc130 = 130,
Ccc133 = 132,
Ccc14 = 14,
Ccc15 = 15,
Ccc16 = 16,
Ccc17 = 17,
Ccc18 = 18,
Ccc19 = 19,
Ccc20 = 20,
Ccc21 = 21,
Ccc22 = 22,
Ccc23 = 23,
Ccc24 = 24,
Ccc25 = 25,
Ccc26 = 26,
Ccc27 = 27,
Ccc28 = 28,
Ccc29 = 29,
Ccc30 = 30,
Ccc31 = 31,
Ccc32 = 32,
Ccc33 = 33,
Ccc34 = 34,
Ccc35 = 35,
Ccc36 = 36,
Ccc84 = 84,
Ccc91 = 91,
```


#### New Type: HarfBuzzSharp.BufferDiffFlags

```csharp
[Serializable]
public enum BufferDiffFlags {
	ClusterMismatch = 32,
	CodepointMismatch = 16,
	ContentTypeMismatch = 1,
	DottedCirclePresent = 8,
	Equal = 0,
	GlyphFlagsMismatch = 64,
	LengthMismatch = 2,
	NotdefPresent = 4,
	PositionMismatch = 128,
}
```

#### New Type: HarfBuzzSharp.OpenTypeColorLayer

```csharp
public struct OpenTypeColorLayer, System.IEquatable<OpenTypeColorLayer> {
	// properties
	public uint ColorIndex { get; set; }
	public uint Glyph { get; set; }
	// methods
	public virtual bool Equals (OpenTypeColorLayer obj);
	public override bool Equals (object obj);
	public override int GetHashCode ();
	public static bool op_Equality (OpenTypeColorLayer left, OpenTypeColorLayer right);
	public static bool op_Inequality (OpenTypeColorLayer left, OpenTypeColorLayer right);
}
```

#### New Type: HarfBuzzSharp.OpenTypeColorPaletteFlags

```csharp
[Serializable]
public enum OpenTypeColorPaletteFlags {
	Default = 0,
	UsableWithDarkBackground = 2,
	UsableWithLightBackground = 1,
}
```

#### New Type: HarfBuzzSharp.OpenTypeLayoutBaselineTag

```csharp
[Serializable]
public enum OpenTypeLayoutBaselineTag {
	Hanging = 1751215719,
	HbOtLayoutBaselineTagMaxValue = 2147483647,
	IdeoEmboxBottomOrLeft = 1768187247,
	IdeoEmboxTopOrRight = 1768191088,
	IdeoFaceBottomOrLeft = 1768121954,
	IdeoFaceTopOrRight = 1768121972,
	Math = 1835103336,
	Roman = 1919905134,
}
```

#### New Type: HarfBuzzSharp.OpenTypeLayoutGlyphClass

```csharp
[Serializable]
public enum OpenTypeLayoutGlyphClass {
	BaseGlyph = 1,
	Component = 4,
	Ligature = 2,
	Mark = 3,
	Unclassified = 0,
}
```

#### New Type: HarfBuzzSharp.OpenTypeMathConstant

```csharp
[Serializable]
public enum OpenTypeMathConstant {
	AccentBaseHeight = 6,
	AxisHeight = 5,
	DelimitedSubFormulaMinHeight = 2,
	DisplayOperatorMinHeight = 3,
	FlattenedAccentBaseHeight = 7,
	FractionDenomDisplayStyleGapMin = 40,
	FractionDenominatorDisplayStyleShiftDown = 35,
	FractionDenominatorGapMin = 39,
	FractionDenominatorShiftDown = 34,
	FractionNumDisplayStyleGapMin = 37,
	FractionNumeratorDisplayStyleShiftUp = 33,
	FractionNumeratorGapMin = 36,
	FractionNumeratorShiftUp = 32,
	FractionRuleThickness = 38,
	LowerLimitBaselineDropMin = 21,
	LowerLimitGapMin = 20,
	MathLeading = 4,
	OverbarExtraAscender = 45,
	OverbarRuleThickness = 44,
	OverbarVerticalGap = 43,
	RadicalDegreeBottomRaisePercent = 55,
	RadicalDisplayStyleVerticalGap = 50,
	RadicalExtraAscender = 52,
	RadicalKernAfterDegree = 54,
	RadicalKernBeforeDegree = 53,
	RadicalRuleThickness = 51,
	RadicalVerticalGap = 49,
	ScriptPercentScaleDown = 0,
	ScriptScriptPercentScaleDown = 1,
	SkewedFractionHorizontalGap = 41,
	SkewedFractionVerticalGap = 42,
	SpaceAfterScript = 17,
	StackBottomDisplayStyleShiftDown = 25,
	StackBottomShiftDown = 24,
	StackDisplayStyleGapMin = 27,
	StackGapMin = 26,
	StackTopDisplayStyleShiftUp = 23,
	StackTopShiftUp = 22,
	StretchStackBottomShiftDown = 29,
	StretchStackGapAboveMin = 30,
	StretchStackGapBelowMin = 31,
	StretchStackTopShiftUp = 28,
	SubSuperscriptGapMin = 15,
	SubscriptBaselineDropMin = 10,
	SubscriptShiftDown = 8,
	SubscriptTopMax = 9,
	SuperscriptBaselineDropMax = 14,
	SuperscriptBottomMaxWithSubscript = 16,
	SuperscriptBottomMin = 13,
	SuperscriptShiftUp = 11,
	SuperscriptShiftUpCramped = 12,
	UnderbarExtraDescender = 48,
	UnderbarRuleThickness = 47,
	UnderbarVerticalGap = 46,
	UpperLimitBaselineRiseMin = 19,
	UpperLimitGapMin = 18,
}
```

#### New Type: HarfBuzzSharp.OpenTypeMathGlyphPart

```csharp
public struct OpenTypeMathGlyphPart, System.IEquatable<OpenTypeMathGlyphPart> {
	// properties
	public int EndConnectorLength { get; set; }
	public OpenTypeMathGlyphPartFlags Flags { get; set; }
	public int FullAdvance { get; set; }
	public uint Glyph { get; set; }
	public int StartConnectorLength { get; set; }
	// methods
	public virtual bool Equals (OpenTypeMathGlyphPart obj);
	public override bool Equals (object obj);
	public override int GetHashCode ();
	public static bool op_Equality (OpenTypeMathGlyphPart left, OpenTypeMathGlyphPart right);
	public static bool op_Inequality (OpenTypeMathGlyphPart left, OpenTypeMathGlyphPart right);
}
```

#### New Type: HarfBuzzSharp.OpenTypeMathGlyphPartFlags

```csharp
[Serializable]
public enum OpenTypeMathGlyphPartFlags {
	Extender = 1,
}
```

#### New Type: HarfBuzzSharp.OpenTypeMathGlyphVariant

```csharp
public struct OpenTypeMathGlyphVariant, System.IEquatable<OpenTypeMathGlyphVariant> {
	// properties
	public int Advance { get; set; }
	public uint Glyph { get; set; }
	// methods
	public virtual bool Equals (OpenTypeMathGlyphVariant obj);
	public override bool Equals (object obj);
	public override int GetHashCode ();
	public static bool op_Equality (OpenTypeMathGlyphVariant left, OpenTypeMathGlyphVariant right);
	public static bool op_Inequality (OpenTypeMathGlyphVariant left, OpenTypeMathGlyphVariant right);
}
```

#### New Type: HarfBuzzSharp.OpenTypeMathKern

```csharp
[Serializable]
public enum OpenTypeMathKern {
	BottomLeft = 3,
	BottomRight = 2,
	TopLeft = 1,
	TopRight = 0,
}
```

#### New Type: HarfBuzzSharp.OpenTypeMetaTag

```csharp
[Serializable]
public enum OpenTypeMetaTag {
	DesignLanguages = 1684827751,
	HbOtMetaTagMaxValue = 2147483647,
	SupportedLanguages = 1936485991,
}
```

#### New Type: HarfBuzzSharp.OpenTypeNameEntry

```csharp
public struct OpenTypeNameEntry, System.IEquatable<OpenTypeNameEntry> {
	// properties
	public IntPtr Language { get; set; }
	public OpenTypeNameId NameId { get; set; }
	public int Var { get; set; }
	// methods
	public virtual bool Equals (OpenTypeNameEntry obj);
	public override bool Equals (object obj);
	public override int GetHashCode ();
	public static bool op_Equality (OpenTypeNameEntry left, OpenTypeNameEntry right);
	public static bool op_Inequality (OpenTypeNameEntry left, OpenTypeNameEntry right);
}
```

#### New Type: HarfBuzzSharp.OpenTypeNameId

```csharp
[Serializable]
public enum OpenTypeNameId {
	CID_FINDFONT_NAME = 20,
	COPYRIGHT = 0,
	DARK_BACKGROUND = 24,
	DESCRIPTION = 10,
	DESIGNER = 9,
	DESIGNER_URL = 12,
	FONT_FAMILY = 1,
	FONT_SUBFAMILY = 2,
	FULL_NAME = 4,
	INVALID = 65535,
	LICENSE = 13,
	LICENSE_URL = 14,
	LIGHT_BACKGROUND = 23,
	MAC_FULL_NAME = 18,
	MANUFACTURER = 8,
	POSTSCRIPT_NAME = 6,
	SAMPLE_TEXT = 19,
	TRADEMARK = 7,
	TYPOGRAPHIC_FAMILY = 16,
	TYPOGRAPHIC_SUBFAMILY = 17,
	UNIQUE_ID = 3,
	VARIATIONS_PS_PREFIX = 25,
	VENDOR_URL = 11,
	VERSION_STRING = 5,
	WWS_FAMILY = 21,
	WWS_SUBFAMILY = 22,
}
```

#### New Type: HarfBuzzSharp.OpenTypeVarAxis

```csharp
public struct OpenTypeVarAxis, System.IEquatable<OpenTypeVarAxis> {
	// properties
	public float DefaultValue { get; set; }
	public float MaxValue { get; set; }
	public float MinValue { get; set; }
	public OpenTypeNameId NameId { get; set; }
	public uint Tag { get; set; }
	// methods
	public virtual bool Equals (OpenTypeVarAxis obj);
	public override bool Equals (object obj);
	public override int GetHashCode ();
	public static bool op_Equality (OpenTypeVarAxis left, OpenTypeVarAxis right);
	public static bool op_Inequality (OpenTypeVarAxis left, OpenTypeVarAxis right);
}
```

#### New Type: HarfBuzzSharp.OpenTypeVarAxisFlags

```csharp
[Serializable]
public enum OpenTypeVarAxisFlags {
	HbOtVarAxisFlagMaxValue = 2147483647,
	Hidden = 1,
}
```

#### New Type: HarfBuzzSharp.OpenTypeVarAxisInfo

```csharp
public struct OpenTypeVarAxisInfo, System.IEquatable<OpenTypeVarAxisInfo> {
	// properties
	public uint AxisIndex { get; set; }
	public float DefaultValue { get; set; }
	public OpenTypeVarAxisFlags Flags { get; set; }
	public float MaxValue { get; set; }
	public float MinValue { get; set; }
	public OpenTypeNameId NameId { get; set; }
	public uint Tag { get; set; }
	// methods
	public virtual bool Equals (OpenTypeVarAxisInfo obj);
	public override bool Equals (object obj);
	public override int GetHashCode ();
	public static bool op_Equality (OpenTypeVarAxisInfo left, OpenTypeVarAxisInfo right);
	public static bool op_Inequality (OpenTypeVarAxisInfo left, OpenTypeVarAxisInfo right);
}
```

#### New Type: HarfBuzzSharp.Variation

```csharp
public struct Variation, System.IEquatable<Variation> {
	// properties
	public uint Tag { get; set; }
	public float Value { get; set; }
	// methods
	public virtual bool Equals (Variation obj);
	public override bool Equals (object obj);
	public override int GetHashCode ();
	public static bool op_Equality (Variation left, Variation right);
	public static bool op_Inequality (Variation left, Variation right);
}
```

mattleibow commented:
I know about the enums and the 2 vars on glyph info.

Can you confirm that all these others are good too?

Gillibald commented:
There are multiple new types that were previously not exposed. There were plans to expose them in the future so it is fine to introduce them now. We just have to make sure they are accessible.

Gillibald commented:
Will have a deeper look shortly.

Gillibald commented:
Maybe we can just generate what we need to cover the current API and add everything else for the `2.7.x` version of this library?

mattleibow commented:
@Gillibald Sounds like a good plan. Was thinking that too.

And then with v2.7.x, we can make some more changes - for example those spans and any parameter/property types.
